### PR TITLE
Optimize RTP header extension handling

### PR DESCRIPTION
**Pull request by jmillan**
Yet another extract from `memory-optimizations` branch.

Use `std::array` instead of a  map for one byte extensions.

### CODE REVIEW

File: worker/include/RTC/RtpPacket.hpp [hunk 1]
```diff
@@ -473,9 +474,11 @@ namespace RTC
 			}
 			else if (HasOneByteExtensions())
 			{
-				auto it = this->mapOneByteExtensions.find(id);
+				if (id > 14)
+					return false;
 
-				return it != this->mapOneByteExtensions.end();
+				// `-1` because we have 14 elements total 0..=13 and `id` is in the range 1..=14
```
Comment (ibc) [inline worker/include/RTC/RtpPacket.hpp]:
```suggestion
				// `-1` because we have 14 elements total 0..=13 and `id` is in the range 1..=14.
```

BTW what does `0..=13` notation men?

File: worker/include/RTC/RtpPacket.hpp [hunk 2]
```diff
@@ -508,12 +511,14 @@ namespace RTC
...
 			else if (HasOneByteExtensions())
 			{
-				auto it = this->mapOneByteExtensions.find(id);
-
-				if (it == this->mapOneByteExtensions.end())
+				if (id > 14)
 					return nullptr;
 
-				auto* extension = it->second;
+				// `-1` because we have 14 elements total 0..=13 and `id` is in the range 1..=14
```
Comment (ibc) [inline worker/include/RTC/RtpPacket.hpp]: Same.

File: worker/include/RTC/SenderBandwidthEstimator.hpp [hunk 3]
```diff
@@ -6,6 +6,7 @@
 #include "RTC/RateCalculator.hpp"
 #include "RTC/SeqManager.hpp"
 #include "RTC/TrendCalculator.hpp"
+#include <absl/container/btree_map.h>
```
Comment (ibc) [inline worker/include/RTC/SenderBandwidthEstimator.hpp:9]: Why is this included her if not used in this .hpp file? If just used in the .cpp file then it must be included there.

File: worker/src/RTC/RtpPacket.cpp [hunk 4]
```diff
@@ -491,8 +489,9 @@ namespace RTC
 				if (extension.id == 0 || extension.id > 14 || extension.len == 0 || extension.len > 16)
 					continue;
 
-				// Store the One-Byte extension element in the map.
-				this->mapOneByteExtensions[extension.id] = reinterpret_cast<OneByteExtension*>(ptr);
+				// Store the One-Byte extension element in an array.
+				// `-1` because we have 14 elements total 0..=13 and `id` is in the range 1..=14
```
Comment (ibc) [inline worker/src/RTC/RtpPacket.cpp]: Same.

File: worker/src/RTC/RtpPacket.cpp [hunk 5]
```diff
@@ -575,12 +574,12 @@ namespace RTC
 		}
 		else if (HasOneByteExtensions())
 		{
-			auto it = this->mapOneByteExtensions.find(id);
+			// `-1` because we have 14 elements total 0..=13 and `id` is in the range 1..=14
```
Comment (ibc) [inline worker/src/RTC/RtpPacket.cpp]: Same

File: worker/src/RTC/RtpPacket.cpp [hunk 6]
```diff
@@ -885,8 +884,9 @@ namespace RTC
 						break;
 					}
 
-					// Store the One-Byte extension element in the map.
-					this->mapOneByteExtensions[id] = reinterpret_cast<OneByteExtension*>(ptr);
+					// Store the One-Byte extension element in an array.
+					// `-1` because we have 14 elements total 0..=13 and `id` is in the range 1..=14
```
Comment (ibc) [inline worker/src/RTC/RtpPacket.cpp]: Same

File: worker/include/RTC/SenderBandwidthEstimator.hpp (hunk 3 above)
Comment (jmillan) [reply worker/include/RTC/SenderBandwidthEstimator.hpp:9]: It is used in the .hpp file, but it was taken from `RtpPacket.hpp` before.

**Review by ibc — APPROVED** [review]
I don really know how that lib works but since tests pass I assume it's ok

File: worker/include/RTC/RtpPacket.hpp (hunk 1 above)
Comment (nazar-pc) [reply worker/include/RTC/RtpPacket.hpp]: This is Rust syntax for ranges, `=` before second number means that that number is included in the range. This is just to avoid ambiguity.

Comment (ibc) [reply worker/include/RTC/RtpPacket.hpp]: 👍🏼

**Review by nazar-pc — APPROVED** [review]
The first observation here is that for 1-byte extensions there is a limit of 14, so it is way cheaper to just allocate contiguous space for 14 bytes as a static array rather than using red-black tree-based map from standard library. This provides huge savings on trivial temporary heap allocations.

For two-byte header extensions just more efficient data structure is used.

File: worker/include/RTC/RtpPacket.hpp [hunk 7]
```diff
@@ -611,8 +616,10 @@ namespace RTC
 		Header* header{ nullptr };
 		uint8_t* csrcList{ nullptr };
 		HeaderExtension* headerExtension{ nullptr };
-		absl::btree_map<uint8_t, OneByteExtension*> mapOneByteExtensions;
-		absl::btree_map<uint8_t, TwoBytesExtension*> mapTwoBytesExtensions;
+		// There might be up to one-byte header extensions
```
Comment (nazar-pc) [inline worker/include/RTC/RtpPacket.hpp]:
```suggestion
		// There might be up to 14 one-byte header extensions
```
